pydantic/_internal/_model_construction.py

```diff
@@ -266,6 +266,17 @@ def init_private_attributes(self: BaseModel, __context: Any) -> None:
     object_setattr(self, '__pydantic_private__', pydantic_private)
 
 
+def get_model_post_init(namespace: dict[str, Any], bases: tuple[type[Any], ...]) -> Callable[..., Any] | None:
+    """Get the `model_post_init` method from the namespace or the class bases, or `None` if not defined."""
+    if 'model_post_init' in namespace:
+        return namespace['model_post_init']
+
+    for base in bases:
+        model_post_init = getattr(base, 'model_post_init', None)
```

Comment:
This will actually always be defined for any base, as any subclass of `BaseModel` will inherit from the original empty method:

https://github.com/pydantic/pydantic/blob/c987bc6398b898d5c9e49606e3acb17b732359e8/pydantic/main.py#L417-L421

So if we have the following:

```python
from pydantic import BaseModel

class A(BaseModel):
    a: int = 1

class B(BaseModel):
    b: int = 1

    def model_post_init(self, __context):
        """Do custom things"""

class C(A, B):
    _private: bool = True
```

`get_model_post_init` will use the empty `model_post_init` available on `A`, and skip the custom one from `B`. Any ideas on how to prevent this? Maybe a custom flag could help, similar to:

https://github.com/pydantic/pydantic/blob/c987bc6398b898d5c9e49606e3acb17b732359e8/pydantic/main.py#L167-L168

?

Reply:
If your concern is to make sure attribute lookup only considers things defined on the exact class, the way we achieve that elsewhere is to look in `base.__dict__`. So, if you do `if 'model_post_init' in base.__dict__`, that tells you if it was defined specifically in `base` vs. in one of its subclasses. The best way to do it might be to loop over the `__mro__` for the class looking for items in the classes' `__dict__`s.

Thinking this through, I think if we want to auto-generate a method but not have that auto-generated version factor into inheritance, using a flag like you've described is probably the best way. _But_, since in this case the method _is_ defined on BaseModel, I think the correct behavior would actually involve calling the BaseModel version in your code above, since that is consistent with standard python behavior for method resolution:

```python
class A:
    def my_method(self):
        print('A')

class B(A):
    pass

class C:
    def my_method(self):
        print('C')

class D(B, C):
    pass

D().my_method()
#> A
```

Not sure if I'm overlooking something important here; in general I'm open to other points of view but I think it will be important that whatever we do doesn't introduce breaking changes here. More generally, I won't be surprised if the best solution ends up being for users to just be more proactive about overriding `model_post_init` in subclasses when inheriting from multiple base models and not being able to put the one with the desired `model_post_init` first in the inheritance order.

Reply:
Also, I think if all you are wanting to do is specifically ignore `BaseModel.model_post_init`, it might be best to add something like:
```python
if model_post_init is BaseModel.model_post_init:
    continue
```